pkg/controller/nodemaintenance/master_quorum_check.go

```diff
@@ -0,0 +1,98 @@
+package nodemaintenance
+
+import (
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubernetes "k8s.io/client-go/kubernetes"
+)
+
+const OpenshiftMachineConfigNamespace = "openshift-machine-config-operator"
+
+func checkValidQuorum(client kubernetes.Interface, node *corev1.Node) (bool, bool, error) {
+	if isMasterNode(client, node) {
+		isFallbackCheck := false
+		validQuorum, err := checkValidQuorumMcoPdb(client)
+		if err != nil {
+			log.Infof("mco pdb check failed. use fallback check for quorum validity. error: %v", err)
+			isFallbackCheck = true
+			// as a fallback if machine config operator is not available:
+			// compute the quorum validity based on current number of master nodes.
+			validQuorum, err = checkQuorumValidityFallback(client, node.Name)
+			if err != nil {
+				err = fmt.Errorf("Fallback; can't check quorum validity %v", err)
+				log.Error(err)
+				return false, isFallbackCheck, err
+			}
+		}
+		log.Infof("node %s is a master. quorum status: %t", node.Name, validQuorum)
+		return validQuorum, isFallbackCheck, nil
+	} else {
+		log.Infof("node %s is not a master", node.Name)
+	}
+	return true, false, nil
+}
+
+func checkValidQuorumMcoPdb(client kubernetes.Interface) (bool, error) {
+
+	pdbList, err := client.Policy().PodDisruptionBudgets(OpenshiftMachineConfigNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return false, fmt.Errorf("can't get mco pdb: %v", err)
+	}
+	if pdbList.Size() == 0 {
+		return false, fmt.Errorf("mco pdb is empty")
+	}
+	log.Infof("pdbListSize %d", pdbList.Size())
+
+	minAvailable := int32(pdbList.Items[0].Spec.MinAvailable.IntValue())
+	currentHealthy := pdbList.Items[0].Status.CurrentHealthy
+	disruptedPodLen := int32(len(pdbList.Items[0].Status.DisruptedPods))
+	isValidQuorum := minAvailable <= currentHealthy-disruptedPodLen-1
+
+	log.Infof("master quorum isvalid: %t machine configuration pdb: spec.minAvailable: %d current healthy: %d disruptedPods %d", isValidQuorum, minAvailable, currentHealthy, disruptedPodLen)
+
+	return isValidQuorum, nil
+}
+
+func checkQuorumValidityFallback(client kubernetes.Interface, nodeName string) (bool, error) {
+
+	masterNodes, availableMasterNodes, err := countMasterNodes(client)
+	if err != nil {
+		return false, err
+	}
+
+	quorumValid := availableMasterNodes-1 >= (masterNodes/2 + 1)
+
+	log.Infof("Node %s availableMasters %d totalMasters %d quorumValid %t", nodeName, availableMasterNodes, masterNodes, quorumValid)
+
+	return quorumValid, nil
+}
+
+func countMasterNodes(client kubernetes.Interface) (int, int, error) {
+
+	masterNodes := 0
+	availableMasterNodes := 0
+
+	n, err := client.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: MasterNodeLabelSelector})
+	if err != nil {
+		return -1, -1, err
+	}
+	for _, node := range n.Items {
+		if !node.Spec.Unschedulable {
```

Comment:
I believe Mike L noticed that not all failed nodes get marked as `Unschedulable`.
You might need to find something in node.Status that is a better indicator.

Reply:
one possibility is to check if kubelet has updated the lease for that node recently, but that won't kick in for quite some time., until kubelet rececks the lease that is, and it is probably not very reliable when there is a large number of nodes.

Reply:
or we just might issue a warning and not check the condition if there is no PDB; so that there is no fallback option. Is that preferable?

Reply:
Maybe it's enough in any case to check node.Status.Phase == v1.NodeTerminated

However right now I don't see where this field value is set in kubernetes.

./pkg/apis/core/types.go
	// NodeTerminated means the node has been removed from the cluster.
	NodeTerminated NodePhase = "Terminated"

Reply:
https://github.com/kubernetes/kubernetes/pull/30005/files suggests that's not a good idea:

> The field is never populated, and now is deprecated.

You could either try to infer how many are healthy by looking at any reason the node might not hosting Pods (ie. look at taints too), or re-implement PDB by counting the number of etcd pods.

Probably better to do neither and just create a PDB in our own namespace (with the same definition) if an official one doesn't exist.

Reply:
pushed the taint based solution.

Reply:
pushed fallback solution based on pdb.